src/adapter/z-stack/adapter/zStackAdapter.ts

```diff
@@ -1117,6 +1096,27 @@ class ZStackAdapter extends Adapter {
         });
     }
 
+    private waitForAreqZdo(command: string, payload?: ZpiObjectPayload): {start: () => Promise<ZpiObject>; ID: number} {
```

Comment:
The status checking is implemented directly in the Zdo spec with `BuffaloZdo.readResponse(clusterId, buffer);` [link](https://github.com/Koenkk/zigbee-herdsman/blob/557ec54b89c1ec6f6a209726473267163bd44ee3/src/zspec/zdo/buffaloZdo.ts#L1420), so you could remove some of the logic here, but I guess that would mean bypassing the ZpiObject parsing for AREQ/ZDO and using the spec instead?
(it would automatically add support for r23 too, since the spec **should be** r23 compliant)

Reply:
> but I guess that would mean bypassing the ZpiObject parsing for AREQ/ZDO and using the spec instead?

The difference with ember is that there is one ZDO response callback where you get the full buffer, for zstack there is a callback per ZDO type, so we do not have the full buffer. So I don't think we can use this for zstack (?)

Reply:
Seems like zstack is modding it (like not passing the transaction seq num?), so that wouldn't work as-is. 😢
If it's just the seq num, I guess we could make a param in the spec to not skip 1 byte when reading?

Reply:
So it means we need to reconstruct the buffer again from the parsed ZDO message?

Reply:
Glanced at the code... I think something like this should allow a smooth-ish transition.

in ZpiObject:

```typescript
const ZDO_RSP_COMMAND_ID_TO_CLUSTER_ID: Map<number, Zdo.ClusterId> = new Map([
    [128, Zdo.ClusterId.NETWORK_ADDRESS_RESPONSE],
    // etc.
]);
```

```typescript
    public static fromUnpiFrame(frame: UnpiFrame): ZpiObject {
        if (frame.type === Type.AREQ && frame.subsystem === Subsystem.ZDO) {
            const clusterId = ZDO_RSP_COMMAND_ID_TO_CLUSTER_ID.get(frame.commandID);

            return new ZpiObject(
                frame.type,
                frame.subsystem,
                Zdo.ClusterId[clusterId],
                frame.commandID,// or clusterId?
                BuffaloZdo.readResponse(clusterId, frame.data, false),
                null,// will this be ok? (avoids lookup)
            );
        }

       // ...
```

in BuffaloZdo:
```typescript
    public static readResponse(clusterId: ZdoClusterId, buffer: Buffer, hasOverhead: boolean = true): unknown {
        const buffalo = new BuffaloZdo(buffer, hasOverhead ? ZDO_MESSAGE_OVERHEAD : 0); // set pos to skip `transaction sequence number`
       // ...
```

This would require matching the Zdo spec payloads throughout the code (object keys, [all typed though](https://github.com/Koenkk/zigbee-herdsman/blob/557ec54b89c1ec6f6a209726473267163bd44ee3/src/zspec/zdo/definition/tstypes.ts#L225)) and likely some tweaks to response waiters (cluster name instead of command name?). And handling the ZDO status throwing where `fromUnpiFrame` is called.